include/deal.II/numerics/smoothness_estimator.h

```diff
@@ -0,0 +1,268 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_smoothness_estimator_h
+#define dealii_smoothness_estimator_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/fe/fe_series.h>
+
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * A namespace for various smoothness estimation strategies for hp-adaptive FEM.
+ */
+namespace SmoothnessEstimator
+{
+  /**
+   * Estimate smoothness from decay of Legendre absolute values of coefficients
+   * on the reference cell.
+   *
+   * In one dimension, the finite element solution on the reference element with
+   * polynomial degree $p$ can be written as
+   * @f[
+   *    u(x) = \sum_{j=0}^{p} a_j P_j(x)
+   * @f]
+   * where $\{P_j(x)\}$ are Legendre polynomials. The decay of the coefficients
+   * is estimated by performing the least squares fit of
+   * @f[
+   *   |a_j| \sim C exp(-\sigma j)
+   * @f]
+   * for $j=1,..,p$. The rate of the decay can be used to estimate the
+   * smoothness. For example, p-refinement will be performed if $\sigma>1$.
+   *
+   * Extension to higher dimension is done by performing the fit in each
+   * direction separately and then taking the lowest value of $\sigma$.
+   *
+   * The returned value for each vector and for each cell is $\exp(-\sigma)$.
+   *
+   * @param fe_series FESeries::Legendre object to calculate coefficients. This
+   * object needs to be initialized to have at least $p+1$ coefficients in each
+   * direction, where $p$ is the maximum polynomial degree to be used.
+   * @param dof_hander an hp::DoFHandler
+   * @param all_solutions a vector of pointers to the solution vectors
+   * @param all_smoothness_indicators a vector of pointers to the smoothness indicators for each @p all_solutions.
+   * @param smallest_abs_coefficient is the smallest absolute value of the
+   * coefficient to be used in linear regression. In case there are less than
+   * two non-zero coefficients, the direction will be skipped. If all
+   * coefficients are zero, the returned value for this cell will be zero (i.e.
+   * smooth).
```

Comment:
what does "i.e., smooth" mean?

Reply:
I have to admit that I don't understand what this parameter does. You say it's the "smallest value to be used in linear regression", but a linear regression doesn't actually *use* values -- it's just computes them. Can you clarify?

Reply:
> You say it's the "smallest value to be used in linear regression"

yes, it's the smallest allowed coefficient to be used in linear regression. Recall that these are Legendre coefficients and they can easily be zero (some of them), i.e. due to symmetry of the FE solution on the element. So this check prevents using such coefficients in linear regression.

p.s. we may end up having coefficients:  `x x 0 x 0 x 0 x 0`, where `x` are non-zero and `0` are zeros.

Reply:
Oh, you're saying that if a computed coefficient is 1e-9 and the tolerance is set to 1e-8, then you'll just ignore the 1e-9 and do the fit based on only the others? I think it would be good to put essentially what you explained above into the documentation, including the example with the `x x 0 ...`